contracts/contracts/GrantRegistry.sol

```diff
@@ -1,135 +1,124 @@
-    return grants[_id].state == State.CONFIG;
+  // --- Core methods ---
+  /**
+   * @notice Create a new grant in the registry
+   * @param _owner Grant owner (has permissions to modify grant information)
+   * @param _payee Address that receives funds donated to this grant
+   * @param _metaPtr URL pointing to grant metadata (for off-chain use)
+   */
+  function createGrant(
+    address _owner,
+    address _payee,
+    string memory _metaPtr
+  ) external {
+    uint96 _id = grantCount;
+    grants[_id] = Grant(_id, _owner, _payee, _metaPtr);
+    emit GrantCreated(_id, _owner, _payee, _metaPtr);
+    grantCount += 1;
   }
 
-  function isActive(bytes32 _id) public view returns (bool) {
-    return grants[_id].state == State.ACTIVE;
+  /**
+   * @notice Update the owner of a grant
+   * @param _id ID of grant to update
+   * @param _owner New owner address
+   */
+  function updateGrantOwner(uint96 _id, address _owner) external {
+    Grant storage grant = grants[_id];
+    require(msg.sender == grant.owner, "Not authorized");
+    grant.owner = _owner;
+    emit GrantUpdated(grant.id, grant.owner, grant.payee, grant.metaPtr);
   }
 
-  function isClosed(bytes32 _id) public view returns (bool) {
-    return grants[_id].state == State.CLOSED;
+  /**
+   * @notice Update the payee of a grant
+   * @param _id ID of grant to update
+   * @param _payee New payee address
+   */
+  function updateGrantPayee(uint96 _id, address _payee) external {
+    Grant storage grant = grants[_id];
+    require(msg.sender == grant.owner, "Not authorized");
+    grant.payee = _payee;
+    emit GrantUpdated(grant.id, grant.owner, grant.payee, grant.metaPtr);
   }
 
-  function getGrantCount() public view returns (uint256) {
-    return grantCount;
+  /**
+   * @notice Update the payee of a grant
+   * @param _id ID of grant to update
+   * @param _metaPtr New URL that points to grant metadata
+   */
+  function updateGrantMetaPtr(uint96 _id, string calldata _metaPtr) external {
+    Grant storage grant = grants[_id];
+    require(msg.sender == grant.owner, "Not authorized");
+    grant.metaPtr = _metaPtr;
+    emit GrantUpdated(grant.id, grant.owner, grant.payee, grant.metaPtr);
   }
 
-  function getAllGrantIds() public view returns (bytes32[] memory) {
-    bytes32[] memory ret = new bytes32[](grantCount);
-    for (uint256 i = 0; i < grantCount; i++) {
-      ret[i] = grantIds[i];
-    }
-    return grantIds;
+  /**
+   * @notice Update multiple fields of a grant at once
+   * @dev To leave a field unchanged, you must pass in the same value as the current value
+   * @param _id ID of grant to update
+   * @param _owner New owner address
+   * @param _payee New payee address
+   * @param _metaPtr New URL that points to grant metadata
+   */
+  function updateGrant(
+    uint96 _id,
+    address _owner,
+    address _payee,
+    string calldata _metaPtr
+  ) external {
+    Grant storage grant = grants[_id];
```

Comment:
Asking, because I'm not sure off the top of my head: is there any gas difference if we just update the whole grant in place? Thinking it might prevent this storage read.

Reply:
We have to do the storage read to check the owner at the beginning of the function, unless I'm misunderstanding something?

Reply:
Yep, you're right. Reading the `owner` property loads the storage slot for the grant `_id` anyway, and there's no way to avoid that check. I don't think there's any room for improvement here.